src/main/java/com/project/bookapp/controller/CategoryController.java

```diff
@@ -0,0 +1,87 @@
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Category management", description = "Endpoints for categories")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/categories")
+@Validated
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.CREATED)
+    @PostMapping
+    @Operation(summary = "Create a new category",
+            description = "Create a new category")
+    public CategoryDto createCategory(@RequestBody
+                                          @Valid CategoryDto categoryDto) {
+        return categoryService.save(categoryDto);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping
+    @Operation(summary = "Get all categories",
+            description = "Get a list of all available categories")
+    public List<CategoryDto> getAll() {
+        return categoryService.findAll();
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping("/{id}")
+    @Operation(summary = "Get category by id", description = "Get category by certain id")
+    public CategoryDto getCategoryById(@PathVariable Long id) {
+        return categoryService.getById(id);
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PutMapping("/{id}")
+    @Operation(summary = "Update category by id", description = "Update category by certain id")
+    public CategoryDto updateCategory(@PathVariable Long id,
+                                      @RequestBody @Valid CategoryDto categoryDto) {
+        CategoryDto updated = categoryService.update(id, categoryDto);
+        return categoryService.update(id, categoryDto);
+    }
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @DeleteMapping("/{id}")
+    @Operation(summary = "Delete category by id", description = "Delete category by certain id")
+    public void deleteCategory(@PathVariable Long id) {
+        categoryService.deleteById(id);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping("/{id}/books")
+    @Operation(summary = "Get book by category id", description = "Get book by certain category id")
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(@PathVariable Long id) {
+        List<Book> books = bookService.getBooksByCategoryId(id);
+        return books.stream()
+                .map(BookDtoWithoutCategoryIds::new)
+                .collect(Collectors.toList());
```

Comment:
```suggestion
                .toList();
```
Move mapping to the service layer